x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/CopySettingsStep.java

```diff
@@ -0,0 +1,112 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.core.ilm;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.cluster.metadata.MetaData;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.index.Index;
+
+import java.util.Locale;
+import java.util.Objects;
+
+/**
+ * Copy the provided settings from the source to the target index.
+ * <p>
+ * The target index is derived from the source index using the provided prefix.
+ * This is useful for actions like shrink or searchable snapshot that create a new index and migrate the ILM execution from the source
+ * to the target index.
+ */
+public class CopySettingsStep extends ClusterStateActionStep {
+    public static final String NAME = "copy-settings";
+
+    private static final Logger logger = LogManager.getLogger(CopySettingsStep.class);
+
+    private final String[] settingsKeys;
+    private final String indexPrefix;
+
+    public CopySettingsStep(StepKey key, StepKey nextStepKey, String indexPrefix, String... settingsKeys) {
+        super(key, nextStepKey);
+        Objects.requireNonNull(indexPrefix);
+        Objects.requireNonNull(settingsKeys);
+        this.indexPrefix = indexPrefix;
+        this.settingsKeys = settingsKeys;
+    }
+
+    @Override
+    public boolean isRetryable() {
+        return true;
+    }
+
+    public String[] getSettingsKeys() {
+        return settingsKeys;
+    }
+
+    public String getIndexPrefix() {
+        return indexPrefix;
+    }
+
+    @Override
+    public ClusterState performAction(Index index, ClusterState clusterState) {
+        String sourceIndexName = index.getName();
+        IndexMetaData sourceIndexMetadata = clusterState.metaData().index(sourceIndexName);
+        String targetIndexName = indexPrefix + sourceIndexName;
+        IndexMetaData targetIndexMetadata = clusterState.metaData().index(targetIndexName);
+
+        if (sourceIndexMetadata == null) {
+            // Index must have been since deleted, ignore it
+            logger.debug("[{}] lifecycle action for index [{}] executed but index no longer exists", getKey().getAction(), sourceIndexName);
+            return clusterState;
+        }
+
+        if (targetIndexMetadata == null) {
+            String errorMessage = String.format(Locale.ROOT, "index [%s] is being referenced by ILM action [%s] on step [%s] but " +
+                "it doesn't exist", targetIndexName, getKey().getAction(), getKey().getName());
+            logger.debug(errorMessage);
+            throw new IllegalStateException(errorMessage);
+        }
+
+        if (settingsKeys == null || settingsKeys.length == 0) {
+            return clusterState;
+        }
```

Comment:
Super minor, but we can probably stick this before the `targetIndexMetadata == null` check to avoid an error where no settings are copied?